src/newscale/stages.py

```diff
@@ -1,365 +1,497 @@
+
+        :param interface: interface object through which to communicate.
+        :param address: address to communicate to the device on the specified
+            interface. Optional. If unspecified, commands will be passed
+            through directly without trying to select an axis first.
+            
+        .. code-block:: python
+
+            from newscale.interfaces import SerialInterface
+            from newscale.stages import M3LinearSmartStage
+
+            interface = SerialInterface(port='COM4'),
+            stage = M3LinearSmartStage(interface, "01")
+        
+        """
+        # Either address can be passed in XOR interface
+        self.log = logging.getLogger(f"{__name__}.{self.__class__.__name__}")
+        self.address = address
+        self.interface = interface
+        self.log.debug("Establishing host control by requesting firmware"
+                       "version.")
+        self.get_firmware_version()  # Necessary to handshake with the PC.
+        self.set_drive_mode(DriveMode.CLOSED_LOOP)  # Prep stage for abs moves.
+
+    @staticmethod
+    def _get_cmd_str(cmd: Cmd, *args: str):
+        args_str = " " + " ".join([a.upper() for a in args]) if args else ""
+        return f"<{cmd.value.upper()}{args_str}>\r"
+
+    def _parse_reply(self, reply: str):
+        """Turn string reply into a Cmd plus one or more response integers."""
+        reply = parse_stage_reply(reply)
+        # Check for errors here.
+        if len(reply) == 1:
+            if reply[0] == Cmd.ILLEGAL_COMMAND:
+                error_msg = "Prior command cannot be issued in this state."
+                self.log.error(error_msg)
+                raise IllegalCommandError(error_msg)
+            elif reply[0] == Cmd.ILLEGAL_COMMAND_FORMAT:
+                error_msg = "Prior command is formatted incorrectly."
+                self.log.error(error_msg)
+                raise IllegalCommandFormatError(error_msg)
+        return reply
+
+    def _send(self, cmd_str: str):
+        """Send a command and return the parsed reply."""
+        # Note: interface will handle the case where address is None.
+
+        # Note: we must block until we get a reply from the device before
+        # returning to ensure that the axis state has changed. Otherwise,
+        # the device and pc are put into a race condition.
+
+        self.interface.send(cmd_str, address=self.address)
+        return self._parse_reply(self.interface.read(self.address))
+
+    # <01>
+    def get_firmware_version(self):
+        """Get the firmware version.
+
+        Note: this command is issued on startup to establish computer control.
+        To release computer control, issue a <02> command.
+
+        Note: some commands will reply with <24> (ILLEGAL_COMMAND) if host
+        control is not yet established.
+        """
+        return self._send(self._get_cmd_str(Cmd.FIRMWARE_VERSION))[-1]
+
+    # <02>
     def close(self):
```

Comment:
This is to flag that our software is not longer connected to the stage and using it. See command <02> in the [second datasheet](https://github.com/AllenNeuralDynamics/python-newscale/blob/dev/driver_refactor/docs/datasheets/Newscale_PathwaySoftwareManual.pdf).

It's worth mentioning that some commands throw a <24> (Illegal command) unless we first issue a <01> to establish computer control. Our software needs to issue a <02> to clear this flag on the device side; otherwise it will persist as long as the device is powered, making debug sessions very confusing.

Reply:
i see - thanks for explaining!